src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java

```diff
@@ -0,0 +1,426 @@
+
+  private final RecordService recordService;
+  private final SnowflakeTelemetryService telemetryService;
+  private Map<String, String> topic2TableMap;
+
+  // Behavior to be set at the start of connector start. (For tombstone records)
+  private SnowflakeSinkConnectorConfig.BehaviorOnNullValues behaviorOnNullValues;
+
+  // default is true unless the configuration provided is false;
+  // If this is true, we will enable Mbean for required classes and emit JMX metrics for monitoring
+  private boolean enableCustomJMXMonitoring = SnowflakeSinkConnectorConfig.JMX_OPT_DEFAULT;
+
+  // default is at_least_once semantic (To begin with)
+  // TODO: SNOW-526435
+  private SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee ingestionDeliveryGuarantee =
+      SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee.AT_LEAST_ONCE;
+
+  // ------ Streaming Ingest ------ //
+  // needs url, username. p8 key, role name
+  private final SnowflakeStreamingIngestClient streamingIngestClient;
+
+  // Config set in JSON
+  private final Map<String, String> connectorConfig;
+
+  private final String taskId;
+
+  private final String streamingIngestClientName;
+
+  /**
+   * Key is formulated in {@link #partitionChannelKey(String, int)} }
+   *
+   * <p>value is the Streaming Ingest Channel implementation (Wrapped around TopicPartitionChannel)
+   */
+  private final Map<String, TopicPartitionChannel> partitionsToChannel;
+
+  public SnowflakeSinkServiceV2(
+      SnowflakeConnectionService conn, Map<String, String> connectorConfig) {
+    if (conn == null || conn.isClosed()) {
+      throw SnowflakeErrors.ERROR_5010.getException();
+    }
+
+    this.fileSize = SnowflakeSinkConnectorConfig.BUFFER_SIZE_BYTES_DEFAULT;
+    this.recordNum = SnowflakeSinkConnectorConfig.BUFFER_COUNT_RECORDS_DEFAULT;
+    this.flushTime = SnowflakeSinkConnectorConfig.BUFFER_FLUSH_TIME_SEC_DEFAULT;
+    this.conn = conn;
+    this.recordService = new RecordService();
+    this.telemetryService = conn.getTelemetryClient();
+    this.topic2TableMap = new HashMap<>();
+
+    // Setting the default value in constructor
+    // meaning it will not ignore the null values (Tombstone records wont be ignored/filtered)
+    this.behaviorOnNullValues = SnowflakeSinkConnectorConfig.BehaviorOnNullValues.DEFAULT;
+
+    this.connectorConfig = connectorConfig;
+    // Streaming ingest requires a properties.
+    // TODO request a builder pattern for SI Client.
+    Properties streamingClientProps = new Properties();
+    streamingClientProps.putAll(connectorConfig);
+    this.taskId = connectorConfig.getOrDefault(Utils.TASK_ID, "-1");
+    this.streamingIngestClientName =
+        STREAMING_CLIENT_PREFIX_NAME + conn.getConnectorName() + "_" + taskId;
+    this.streamingIngestClient =
+        SnowflakeStreamingIngestClientFactory.builder(this.streamingIngestClientName)
+            .setProperties(streamingClientProps)
+            .build();
+    this.partitionsToChannel = new HashMap<>();
```

Comment:
Doesn't this need to be thread safe?

Reply:
Not needed. Reason being:
1. Each of this instances are only assigned a set of partitions. 
2. Even if there are multiple tasks (SnowflakeSinkTask) assigned to a node (In case of multiple partitions) each of this tasks are only assigned its own set of partitions and no two partitions will overlap across nodes and tasks. 

So we are kind of okay to not make this thread safe. 
Reference:
- https://docs.confluent.io/platform/current/connect/devguide.html#sink-tasks
- https://kafka.apache.org/24/javadoc/org/apache/kafka/connect/sink/SinkTask.html
`Partition Assignment: After initialization, Connect will assign the task a set of partitions using open(Collection). These partitions are owned exclusively by this task until they have been closed with close(Collection).`